File: LeetCode/436-Find Right Interval/solution.py
#!/usr/local/bin/python3
# coding: UTF-8
# Author: David
# Email: youchen.du@gmail.com
# Created: 2016-12-08 19:40
# Last modified: 2016-12-08 19:40
# Filename: solution.py
# Description:
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e
import bisect

class Solution(object):
    def findRightInterval(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[int]
        """
        l = sorted((e.start, i) for i, e in enumerate(intervals))
        res = []
        for k in intervals:
            r = bisect.bisect_left(l, (k.end,))
            res.append(l[r][1] if r < len(l) else -1)
        return res

File: LeetCode/436-Find Right Interval/solution.cpp
// Author: David
// Email: youchen.du@gmail.com
// Created: 2016-12-08 19:35
// Last modified: 2016-12-08 19:35
// Filename: solution.cpp
// Description:
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
 struct CmpItem {
     int value;
     int idx;
     CmpItem(int v, int i): value(v), idx(i) {}
 };
 
 bool cmp(const CmpItem &p, const CmpItem &q)
 {
     return p.value > q.value ? true:false;
 }
 
class Solution {
public:
    vector<int> findRightInterval(vector<Interval>& intervals) {
        int size = intervals.size();
        vector<int> ends, ans;
        vector<CmpItem> starts;
        for(int i=0; i < size; i++)
        {
            ends.push_back(intervals[i].end);
            starts.push_back(CmpItem(intervals[i].start, i));
        }
        sort(starts.begin(), starts.end(), cmp);
        for(int i=0; i < size; i++)
        {
            ans.push_back(binary_search(starts, 0, size, ends[i]));
        }
        return ans;
    }
    
    int binary_search(vector<CmpItem>& starts, int s, int e, int k)
    {
        if(s == e)
        {
            if(s > 0)
                return starts[s-1].idx;
            else
                return -1;
        }
        int m = (s+e)>>1;
        if(starts[m].value == k)
            return starts[m].idx;
        else if(starts[m].value > k)
            return binary_search(starts, m+1, e, k);
        else
            return binary_search(starts, s, m, k);
    }
};
